player/scripting.c

```diff
@@ -266,29 +270,165 @@ void mp_load_builtin_scripts(struct MPContext *mpctx)
+            mp_path_strip_trailing_separator(client_name);
+            client_name = mp_basename(client_name);
+        } else {
+            client_name = script_name_from_filename(tmp, fname);
+        }
+        clients[i] = mp_new_client(mpctx->clients, client_name);
+        // TODO: Check for failure creating clients.
+    }
+    talloc_free(tmp);
+
+    struct mp_script_args *arg = talloc_ptrtype(NULL, arg);
+    *arg = (struct mp_script_args){
+        .mpctx = mpctx,
+        .filename = NULL,
+        .path = NULL,
+        .backend = backend,
+        // Create the client before creating the thread; otherwise a race
+        // condition could happen, where MPContext is destroyed while the
+        // thread tries to create the client.
+        .client = mp_new_client(mpctx->clients, script_name),
+        .clients = clients,
+        .py_scripts = py_scripts,
+        .script_count = script_count,
+    };
+
+    if (!arg->client) {
+        MP_ERR(mpctx, "Failed to create client for python\n");
+        return -1;
+    }
+
+    mp_client_set_weak(arg->client);
+    arg->log = mp_client_get_log(arg->client);
+    int64_t id = mpv_client_id(arg->client);
+
+    MP_DBG(arg, "Loading python scripts...\n");
+    mp_thread thread;
+    // thread detaches itself and never returns
+    mp_thread_create(&thread, script_thread, arg);
+
+    return id;
+}
+
+static void push_to_py_scripts_array(size_t *counts, char **py_scripts, char *file)
+{
+    if (counts[1] == counts[0]) {
+        py_scripts = talloc_realloc(NULL, py_scripts, char *, counts[0] * 2);
+    }
+    py_scripts[counts[0]] = talloc_strdup(NULL, file);
+    counts[0] = counts[0] + 1;
+}
+
+static bool put_in_py_scripts(size_t *counts, char **py_scripts, char *file)
+{
+    char *ext = mp_splitext(file, NULL);
+    if (ext && strcasecmp(ext, "py") == 0) {
+        push_to_py_scripts_array(counts, py_scripts, file);
+        return true;
+    }
+    if (!ext) {
+        // check if it's a directory
+        struct stat s;
+        if (!stat(file, &s) && S_ISDIR(s.st_mode)) {
+            char *filepath = mp_path_join(NULL, file, "__init__.py");
+            if (!stat(filepath, &s) && S_ISREG(s.st_mode)) {
+                push_to_py_scripts_array(counts, py_scripts, file);
+                return true;
+            }
+        }
+    }
+    return false;
+}
+#endif
+
 bool mp_load_scripts(struct MPContext *mpctx)
 {
     bool ok = true;
 
+#if HAVE_PYTHON
+    size_t *py_count = talloc_array(NULL, size_t, 2);
+    py_count[0] = 0; // file count
+    py_count[1] = 2; // array size
+    char **py_scripts = talloc_array(NULL, char *, 2);
+#endif
     // Load scripts from options
     char **files = mpctx->opts->script_files;
     for (int n = 0; files && files[n]; n++) {
-        if (files[n][0])
-            ok &= mp_load_user_script(mpctx, files[n]) >= 0;
+        if (files[n][0]) {
+#if HAVE_PYTHON
+            if (!put_in_py_scripts(py_count, py_scripts, files[n]))
```

Comment:
This file `players/scripting.c` shouldn't reference any specific pything parts. It should be generic. Also we shouldn't keep any `py_scripts` arrays, each client should be separate and spawned when needed. If any bookkeeping is necessary it should be done when such client is created in `load_script` function of specific backend.